### PR TITLE
Allow passing context to ISerializable::deserialize

### PR DESCRIPTION
**Pull request by mbasmanova**
A follow-up PR adds serialization for expressions. To serialize
ConstantTypeExpr, we use VectorSaver to serialize the constant vector. To
deserialize, we need to reconstruct the constant vector, which requires a
memory pool. Hence, we need a way to pass memory pool to
ISerializable::deserialize.

Part of #4303

### CODE REVIEW

File: velox/common/serialization/Serializable.h [hunk 1]
```diff
@@ -179,35 +189,37 @@ class ISerializable {
...
+      void* context) {
+    return T::create(obj, context);
   }
 
   template <
       class T,
       typename =
           std::enable_if_t<std::is_integral_v<T> && !std::is_same_v<T, bool>>>
-  static T deserialize(const folly::dynamic& obj) {
+  static T deserialize(const folly::dynamic& obj, void* context) {
```
Comment (pedroerp) [inline velox/common/serialization/Serializable.h:202]: nit: you need to comment out the parameter name if you're not using it or lint will yell at you `void* /* context */`

File: velox/type/Type.cpp [hunk 2]
```diff
@@ -233,6 +235,14 @@ TypePtr Type::create(const folly::dynamic& obj) {
   }
 }
 
+// static
+void Type::registerSerDe() {
+  velox::DeserializationRegistryForSharedPtr().Register(
```
Comment (pedroerp) [inline velox/type/Type.cpp:240]: nit: why aren't these methods names lower camel case?

File: velox/type/tests/TypeTest.cpp [hunk 3]
```diff
@@ -22,10 +22,7 @@ using namespace facebook::velox;
 
 namespace {
 void testTypeSerde(const TypePtr& type) {
-  velox::DeserializationRegistryForSharedPtr().Register(
-      Type::getClassName(),
-      static_cast<std::shared_ptr<const Type> (*)(const folly::dynamic&)>(
-          Type::create));
+  Type::registerSerDe();
```
Comment (pedroerp) [inline velox/type/tests/TypeTest.cpp:25]: I was expecting here a test that actually passes some context and verifies that you can use while deserializing stuff. Is this coming in the next PR?

**Review by pedroerp — APPROVED** [review]

File: velox/type/Type.cpp (hunk 2 above)
Comment (xiaoxmeng) [reply velox/type/Type.cpp:240]: +1

**Review by xiaoxmeng — APPROVED** [review]
@mbasmanova LGTM. Thanks!

File: velox/type/Type.cpp (hunk 2 above)
Comment (mbasmanova) [reply velox/type/Type.cpp:240]: This is legacy from Koski. We may want to clean this up at some point.

File: velox/type/tests/TypeTest.cpp (hunk 3 above)
Comment (mbasmanova) [reply velox/type/tests/TypeTest.cpp]: This is a "stacked PR" and this change comes from https://github.com/facebookincubator/velox/pull/4306.

I added a unit test for 'context' in deserialize.

File: velox/common/serialization/tests/SerializableTest.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+TEST(SerializableTest, context) {
+  DeserializationWithContextRegistryForSharedPtr().Register("Dot", Dot::create);
+
+  Dot dot{"yellow", 10};
+  auto serialized = dot.serialize();
+
+  // Deserialize with different multipliers for radius passed via 'context'.
+  int32_t multiplier = 1;
+  auto copy = ISerializable::deserialize<Dot>(serialized, &multiplier);
+  ASSERT_EQ(copy->color(), dot.color());
```
Comment (pedroerp) [inline velox/common/serialization/tests/SerializableTest.cpp:79]: don't objects get a default `operator=()` that lets you simply write

```
ASSERT_EQ(copy, dot);
```

?

**Review by pedroerp — APPROVED** [review]
Thanks for the unit tests.

File: velox/common/serialization/tests/SerializableTest.cpp (hunk 4 above)
Comment (mbasmanova) [reply velox/common/serialization/tests/SerializableTest.cpp]: This doesn't work out of the box, but I added operator== and operator* to simplify the test. Thank you for the suggestion.

```
ASSERT_EQ(*copy, dot);
ASSERT_EQ(*copy, dot * 10);
```

**Pull request closed without merging**